docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,210 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+**s2n_config_set_psk_mode** sets the PSK mode on the config object. The PSK mode represents how the pre-shared key was generated and shared between the server and the client. The enum **s2n_psk_mode** lists the supported PSK modes. Use this API prior to the ClientHello to set the PSK mode for the connection. This API can also be used to switch between the different PSK modes.
+
+**s2n_connection_set_psk_mode** sets the PSK mode on the connection object. The enum **s2n_psk_mode** lists the supported PSK modes. This API overrides the PSK mode set on config for this connection. If the connection already contains a list of PSKs **s2n_connection_set_psk_mode** verifies that the current PSK mode matches that of the list. If the PSK mode being set differs from the existing list, an error is returned. If there is no existing list of PSKs present on the connection, the PSK mode of the list is set to the input PSK mode.
+
+### Selecting a Pre-Shared Key
+
+By default, s2n server chooses the first identity in its PSK list that also appears in the client's PSK list. If you would like to implement your own PSK selection logic,
+use the **s2n_psk_selection_callback** to select the PSK to be used for the connection, along with the following offered PSK APIs to process the client sent list of PSKs.
+
+```c
+struct s2n_offered_psk* s2n_offered_psk_new();
+int s2n_offered_psk_free(struct s2n_offered_psk **psk);
+```
+
+**s2n_offered_psk** creates a new offered PSK object. An offered PSK represents a single PSK offered by the client.
+Use this object to create the list of offered PSKs sent by the client. Use **s2n_offered_psk_free** to free
+the memory allocated to the offered PSK object created by this API.
+
+**s2n_offered_psk_free** frees the memory associated with the **s2n_offered_psk** object.
+
+The server calls the following APIs to select the PSK to be used for the connection.
+
+```c
+int s2n_offered_psk_get_identity(struct s2n_offered_psk *psk, uint8_t** identity, uint16_t *size);
+bool s2n_offered_psk_list_has_next(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+int s2n_offered_psk_list_reread(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_choose_psk(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, void *context,
+                                          struct s2n_offered_psk_list *psk_list);
```

Comment:
You should discuss the lifetime / safety of the psk_list variable, either in the doxygen comment or here.

Reply:
The lifetime of the s2n_offered_psk_list is till the end of the PSK selection callback.